bin/init.sh

```diff
@@ -44,51 +69,82 @@ if [ ${ROOT} != "${GO_TOP:-$HOME/go}/src/istio.io/istio" ]; then
        exit 1
 fi
 
-PROXYVERSION=$(grep envoy-debug pilot/docker/Dockerfile.proxy_debug  |cut -d: -f2)
-PROXY=debug-$PROXYVERSION
-
-# Save envoy in $out
-if [ ! -f $OUT/envoy-$PROXYVERSION ] ; then
-    mkdir -p $OUT
-    pushd $OUT
-    # New version of envoy downloaded. Save it to cache, and clean any old version.
+# Download Envoy debug and release binaries for Linux x86_64. They will be included in the
+# docker images created by Dockerfile.proxy and Dockerfile.proxy_debug.
 
-    DOWNLOAD_COMMAND=""
+# Gets the download command supported by the system (currently either curl or wget)
+DOWNLOAD_COMMAND=""
+set_download_command () {
+    # Try curl.
     if command -v curl > /dev/null; then
-       if curl --version | grep Protocols  | grep https > /dev/null; then
-	   DOWNLOAD_COMMAND='curl -fLSs'
-       else
-           echo curl does not support https, will try wget for downloading files.
-       fi
+        if curl --version | grep Protocols  | grep https > /dev/null; then
+	       DOWNLOAD_COMMAND='curl -fLSs'
+	       return
+        fi
+        echo curl does not support https, will try wget for downloading files.
     else
-       echo curl is not installed, will try wget for downloading files.
+        echo curl is not installed, will try wget for downloading files.
     fi
 
-    if [ -z "${DOWNLOAD_COMMAND}" ]; then
-        if command -v wget > /dev/null; then
-	    DOWNLOAD_COMMAND='wget -qO -'
-        else
-            echo wget is not installed.
-        fi
+    # Try wget.
+    if command -v wget > /dev/null; then
+        DOWNLOAD_COMMAND='wget -qO -'
+        return
     fi
+    echo wget is not installed.
+
+    echo Error: curl is not installed or does not support https, wget is not installed. \
+         Cannot download envoy. Please install wget or add support of https to curl.
+    exit 1
+}
+
+# Normally set by the Makefile.
+# Keep the default URLs up-to-date with the latest push from istio/proxy.
+ISTIO_ENVOY_VERSION=${ISTIO_ENVOY_VERSION:-54077a0c8486b6db7b2fc5d10619b6813bb40767}
```

Comment:
I agree with not peppering the code with SHAs and the point about not duplicating makefile logic

can we put at the beginning "if !RAN_THROUGH_MAKE  echo "please type 'make init' to invoke this;" exit 1" equiv

and in Makefile do RAN_THROUGH_MAKE=1 bin/init.sh

Reply:
Updated to use `PROXY_TAG` from `istio.VERSION`.

Reply:
Hrm, in a different PR I came to see that people (and thus the istio source) seem to have different and conflicting impressions on what PROXY_TAG refers to.  Some cases treat this tag as pointing to the envoy images from the proxy repo (e.g., the contents of istio.deps and the files it tells the bots to update), while others cases seem to think this points at the proxy image from the istio repo under pilot/docker (e.g., the -a behavior of install/updateVersion.sh, which sets pilot, proxy, mixer, and ca to all the same values).  So, the first time that someone builds it'll probably do okay, but if someone runs updateVersion.sh (e.g., via "make installgen" or "make push") then I'm expecting things will break because updateVersion.sh will probably update PROXY_TAG in istio.VERSION to point to the istio repo rather than the proxy repo.

Reply:
@mattdelco where do you see PROXY_TAG referencing the pilot Dockerfiles?

Reply:
The PROXY_TAG is referenced in files under install/ and samples/ .  A clearer sign of the conflicting use is:

<pre>samples/bookinfo/eureka/templates/bookinfo.sidecars.yaml.tmpl:    image: {PILOT_HUB}/proxy_debug:{PROXY_TAG}</pre>

Just to be clear, this conflicting use of PROXY_TAG existed prior to your change and the -a behavior is arguably broken and/or there needs to be two sets of proxy tags.  Prior to your change updateVersion.sh would update the PROXY_TAG in istio.VERSION with a istio repo commit hash but I doubt anyone or anything cared.  After your change the build now looks at istio.VERSION so the first "make push" would probably run okay but the second is likely to fail.  The main workaround I can see (to avoid being on the hook to fix this preexisting mess) is to store the commit SHA elsewhere (e.g., the Makefile) and have an entry in istio.deps to update that location.  One of the downsides of both Makefile and istio.deps is they're fairly high-traffic files (with regards to rate of code changes) and it'll be harder to cook up a make rule that efficiently only downloads a new envoy tar when the sha has changed (since these files often change for other reasons, unlike the Dockerfiles).

Reply:
Here's the other PR I had in mind:

https://github.com/istio/istio/pull/3641

IMO it went after the symptom rather than the problem by changing the script to look at istio.deps rather than istio.VERSION (because istio.VERSION is getting changed by updateVersion.sh to point PROXY_TAG at pilot's proxy[_debug] images instead of proxy's envoy images).